### PR TITLE
implement scaffolding for R Notebook filetype registration

### PR DESCRIPTION
**Pull request by kevinushey**
Not ready for merge; pending testing + icons. Note that the work in this PR is modeled off of this commit: https://github.com/rstudio/rstudio/commit/903e86ce1ee4fcddb89c54cb43c8696d82902353
## TODO
- [ ] Icons for the R Notebook format (or just use default HTML icon set provided by OS?)
- [ ] Testing for operating systems:
  - [ ] Windows
  - [ ] OS X
  - [ ] Linux
- [ ] Windows shell associations (for opening in IDE vs. Browser)

### CODE REVIEW

**Comment by jjallaire** [comment]
I'll have Paul do full icon sets for the Notebook icon.

File: package/win32/CMakeLists.txt [hunk 1]
```diff
@@ -34,6 +34,7 @@ set(CPACK_NSIS_EXTRA_INSTALL_COMMANDS "
    WriteRegStr HKCR 'Applications\\\\rstudio.exe\\\\SupportedTypes' '.Rhtml' ''
    WriteRegStr HKCR 'Applications\\\\rstudio.exe\\\\SupportedTypes' '.Rpres' ''
    WriteRegStr HKCR 'Applications\\\\rstudio.exe\\\\SupportedTypes' '.Rproj' ''
+   WriteRegStr HKCR 'Applications\\\\rstudio.exe\\\\SupportedTypes' '.nb.html' ''
```
Comment (jjallaire) [inline package/win32/CMakeLists.txt]: My guess is that this won't work on Windows, let's hope I'm wrong!

File: src/cpp/core/FilePath.cpp [hunk 2]
```diff
@@ -471,6 +471,7 @@ MimeType s_mimeTypes[] =
    { "Rprofile", "text/x-r-source"},
    { "Renviron", "text/x-shell" },
    { "rprofvis",   "text/x-r-profile" },
+   { "nb.html", "text/x-r-notebook" },
```
Comment (jjallaire) [inline src/cpp/core/FilePath.cpp]: I don't think the double-extension will work here either (as we match on just what's after the .nb). I also don't think it's at all problematic to treat the mime type of .nb.html as HTML (will be more broadly compatible).

File: src/cpp/desktop-mac/Info.plist.in [hunk 3]
```diff
@@ -139,6 +139,20 @@
       <dict>
          <key>CFBundleTypeExtensions</key>
          <array>
+             <string>nb.html</string>
+         </array>
+         <key>CFBundleTypeIconFile</key>
+         <string>RNotebook.icns</string>
```
Comment (jjallaire) [inline src/cpp/desktop-mac/Info.plist.in]: I'll get this from Paul.

File: src/gwt/src/org/rstudio/core/client/files/FileSystemItem.java [hunk 4]
```diff
@@ -385,6 +385,7 @@ public final native boolean exists() /*-{
       MIME_TYPES.put( "rmd",   "text/x-r-markdown");
       MIME_TYPES.put( "rhtml", "text/x-r-html");
       MIME_TYPES.put( "rpres", "text/x-r-presentation");
+      MIME_TYPES.put( "nb.html", "text/x-r-notebook");
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/core/client/files/FileSystemItem.java]: Again I don't think it's a problem for .nb.html to have a MIME type of just plain HTML.

File: src/cpp/desktop/resources/freedesktop/rstudio.xml [hunk 5]
```diff
@@ -45,6 +45,12 @@
     <glob pattern="*.rpres"/>
   </mime-type>
 
+  <mime-type type="text/x-r-notebook">
```
Comment (kevinushey) [inline src/cpp/desktop/resources/freedesktop/rstudio.xml]: Does it make sense to define a mime-type here? It looks like this helps RStudio get priority in opening `.nb.html` files + might be needed for icons on Linux; however, we might also want to consult https://specifications.freedesktop.org/shared-mime-info-spec/shared-mime-info-spec-0.18.html to see if there's a better way to declare this.

File: package/win32/CMakeLists.txt (hunk 1 above)
Comment (kevinushey) [inline package/win32/CMakeLists.txt]: It looks like you're right -- it seems like Windows always treats a file extension based on the last `.` in the file name. Was worth a shot!

Comment (jjallaire) [inline package/win32/CMakeLists.txt]: I think the best we can do on Windows without spending an enormous amount of time with shell extensions (which must be implemented in C++ and can't be built with MingGW) is to add an open-with entry for .html, i.e.

```
WriteRegStr HKCR '.html\\\\OpenWithList\\\\rstudio.exe' '' ''
```

Users will then be able to right-click on the notebook to open it in RStudio for editing.

**Comment by jjallaire** [comment]
I think the situation on the Mac is the same (best we can do is have a .html extension).

So I think the only place we can have a file extension that actually works is on Linux. I think that is is okay though -- it is certainly an inconvenience and extra step to require an explicit gesture (either right-click or opening from within RStudio) for opening .nb.html, however this is offset by:

1. The *extreme* convenience on the distribution side of using a .html extension (no viewer needed, works w/ github raw, etc.)

2. Our expectation that developers will collaborate more often using directories (for supporting data, etc.) and side by side .Rmd files that already exist.

One other thing we can do for the use case of single .nb.html file distribution (e.g. tutorials, stand-alone examples, etc.): we can create a gesture off of the **Code** menu in the Rmd to download the Rmd source. This allows the user to view an .nb.html locally in whatever directory they downloaded to (e.g. "Downloads") then explicitly decided to carry on editing their own local version in a path of their choosing.

**Comment by jjallaire** [comment]
Note let's discuss the **Code** menu idea before doing any implementation as I think this is best done as a generic R Markdown feature that .nb.html can just take advantage of.

**Comment by kevinushey** [comment]
I think this PR kind of fell off the radar and we no longer intend on pursuing this, given the difficulty on making extensions of the form `.nb.html` work cross-platform. @jjallaire, should we just close this or are there some parts of this we might still want to implement?

**Comment by jjallaire** [comment]
Yes, I agree we should close this.

**Pull request closed without merging**